Fix serialization P2 issues

jeff303 opened the pull request:
# Serialization P2 Fixes August 2021
## Commit 1: Fix serialization dump error when there are no collections
- Update `select-collections` to correctly handle the case where there are no collections
- Adding new test that removes all collections, then ensures that dump works with no errors

## Commit 2: Fix serialization load error into empty/blank target DB
- Loads dump into completely blank target app DB
- If user is newly created by the load process, they are assigned a random password and a password reset email is generated and sent to them

## Commit 3: Fix serialization objects being incorrectly updated on skip
- Gets rid of unneeded `maybe-fixup-card-template-ids!` fn, which is what was forcing update of some entities even on `:skip`

jeff303 committed:
Fix serialization dump error when there are no collections

Update `select-collections` to correctly handle the case where there are no collections

Adding new test that removes all collections, then ensures that dump works with no errors

jeff303 committed:
Fix serialization load error into empty/blank target DB

Rethrowing exception in cmd when overall load fails

Add test to ensure that a dump containing a user can be loaded into a blank target app DB successfully

Adding a few missing bindings to the `with-temp-empty-app-db` code to set the connection vars under metabase.db.connection

In upsert, add hooks for a :pre-insert-fn and :post-insert-fn to be invoked for new entity instances created by the upsert process, since whether an entity will be an insert or update isn't necessarily known by the load process (only by upsert once identity-condition is checked for each)

In load, setting the pre and post insert functions for a User instance to initialize :password with a random value, and to generate and send a password reset email to the newly inserted user's email, respectively

NOTE: this post insert fn (to send a password reset email for newly inserted users) is NOT hooked up for User as of x.41 release, since it is considered a bugfix, but this can be enabled in a future release)

Adding new defs for the magic permission group names to make those easier to override from tests that might need to (such as the one added for this commit)

jeff303 committed:
Fix serialization objects being incorrectly updated on skip

Remove `maybe-fixup-card-template-ids!`, which was forcing mode :update, since the existing retry logic should cover what it was trying to do

Update Card and Metric models to delete any Dependency instances for which they are the `:model_id` (to make serialization tests after other tests created temporary Card/Dependency pairs)

Adding missing assertions for Dependency serialization

kidd reviewed (approved):
lgtm :+1:

flamber commented on src/metabase/models/metric.clj:
```diff
@@ -17,6 +17,9 @@
 (models/defmodel Metric :metric)
 
+(defn- pre-delete [{:keys [id]}]
+  (db/delete! 'Dependency :model "Metric", :model_id id))
```
Segments are similar to Metrics, but none can be deleted - only retired/archived. Perhaps I'm not understanding what this code is supposed to do.

jeff303 replied:
There likely are other places.  However, the "hard delete" path only seems to be available via tests, which is what led to the error this is intended to fix.  The actual production workflow doesn't do a hard delete, just an archive.  So this change will only impact tests.

flamber reviewed (changes requested):
Testing `ebc9dcd`

When I do a load without any mode-parameter (or mode=skip), then a lot of things stops working. And it still seems to be updating a dashboard, when I re-run the load without mode (#16954)

Without mode=update, it seems like it's breaking 14914 and 14909 and EE533 and EE532 and EE487 - and possibly other nested things?
Also seems to convert some GUI nested questions to SQL - haven't completely understood what's happening here.
Doesn't update the SQL correctly (`-- DUMMY QUERY FOR SERIALIZATION FIRST PASS INSERT`)

When using mode=update, then it seems like it's not correctly removing the existing cards on the dashboard, so all updated cards gets duplicated (I updated the text with "UPDATED" to the two first textboxes, and changed a question to reference another question, and then added a new card)
![image](https://user-images.githubusercontent.com/1447303/129554652-b434ecac-bbbd-4b2c-bb3b-432cd36d643d.png)

I'm just going to block this for now, since it has an approve.

jeff303 commented:
I continue to be troubled at how difficult it is to "see" the same behavior from the backend tests, which are clearly belie the state of the functionality.  Once I narrow down on what is causing this latest round of discrepancies, it might be time to revamp the test suite yet again.

jeff303 commented:
OK, here is the fundamental issue at play, I believe.  It all comes down to the `identity-condition` as seen [here](https://github.com/metabase/metabase/blob/master/enterprise/backend/src/metabase_enterprise/serialization/upsert.clj#L32).  This is the set of fields that will be checked to determine if an entity loaded from the YAML is the same as an existing one in the target app DB.

If the load process reads an entity `A` from YAML, and does not find a "matching" target entity, then `A` will be inserted as a new record into the target DB.  This is true regardless of `--mode`, because the process considers it to be "new".  So in the case of `DashboardCard`, for instance, changing the text in a card will make the load process consider the YAML entity to be new.  That's because the text is technically under the `visualization_settings`, [which **is** part](https://github.com/metabase/metabase/blob/master/enterprise/backend/src/metabase_enterprise/serialization/upsert.clj#L40) of the `identity-condition`.

The `--mode` flag only controls what happens on an identity match (not a miss, which - as outlined above - will always result in an insert). If set to `:update` then the entity will be updated if any other of its fields (i.e. things not in the `identity-condition`) changed.  If not, then it will also be skipped.

One idea would be to simply remove `visualization_settings` from the `identity-condition` for `DashboardCard`, but I don't think that would work.  These text cards only _have_ difference there (the `:card-id` is `nil`).

What we could do is always drop all existing DashboardCard instances `DC` for a Dashboard `D`, if and only if mode is `:update` and we are about to update `D`.  Then all loaded YAML `DC` instances will simply replace everything that was there.  But we will need to think about whether this makes sense, and whether the idea should be extended to other entities.

jeff303 committed:
Serialization: Fix reload entity logic

Now that the `:mode` is always respected instead of being ignored sometimes, we need to update our "second pass" reload functions to always make the mode `:update` on the second pass, or else that entity would just be skipped, which is bad

Also updated the test to use `:mode` `:skip` from the beginning to be more stringent

flamber reviewed (approved):
Tested `3a3e5fb` - it works great :tada: - now :taco: :man_dancing: :wink: after a little lint :broom: 

We probably want to create an issue about https://github.com/metabase/metabase/pull/17388#issuecomment-900415001, so we can figure out how to handle dashcards better (removal, Textbox, multiple of the same card_id, ...)
Do you wanna do it @jeff303, since you got the best technical context?

rlotun commented:
@metabase-bot backport release-x.40.x
